wdl-lint/src/rules/blanks_between_elements.rs

```diff
@@ -0,0 +1,546 @@
+                        state.add(excess_blank_line(p.text_range().to_span()));
+                    }
+                } else {
+                    let first = decl
+                        .syntax()
+                        .prev_sibling()
+                        .is_some_and(|f| f.kind() == SyntaxKind::InputSectionNode);
+
+                    let prev = skip_preceding_comments(decl.syntax());
+
+                    if first {
+                        check_prior_spacing(&prev, state, true, false);
+                    }
+                }
+            }
+        }
+    }
+
+    fn conditional_statement(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        stmt: &wdl_ast::v1::ConditionalStatement,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        let first = stmt
+            .syntax()
+            .prev_sibling()
+            .is_some_and(|f| f.kind() == SyntaxKind::InputSectionNode);
+
+        let prev = skip_preceding_comments(stmt.syntax());
+
+        if first {
+            check_prior_spacing(&prev, state, true, false);
+        }
+    }
+}
+
+/// Check if the given syntax node is the first element in the block.
+fn is_first_element(syntax: &SyntaxNode) -> bool {
+    let mut prev = syntax.prev_sibling_or_token();
+    while let Some(ref cur) = prev {
+        match cur {
+            NodeOrToken::Token(t) => {
+                if t.kind() == SyntaxKind::OpenBrace {
+                    return true;
+                }
+            }
+            NodeOrToken::Node(_) => {
+                return false;
+            }
+        }
+        prev = cur.prev_sibling_or_token();
+    }
+    true
+}
+
+/// Some sections do not allow blank lines, so detect and flag them.
+fn flag_all_blanks(syntax: &SyntaxNode, state: &mut Diagnostics) {
+    syntax.descendants_with_tokens().for_each(|c| {
+        if c.kind() == SyntaxKind::Whitespace {
+            let count = c.to_string().chars().filter(|c| *c == '\n').count();
+            if count > 1 {
+                state.add(excess_blank_line(c.text_range().to_span()));
+            }
+        }
+    });
+}
+
+/// Check that an item has space prior to it.
+/// element_spacing indicates if spacing is required (true) or not (false).
+fn check_prior_spacing(
+    syntax: &NodeOrToken<SyntaxNode, SyntaxToken>,
+    state: &mut Diagnostics,
+    element_spacing: bool,
+    first: bool,
+) {
+    if let Some(prior) = syntax.prev_sibling_or_token() {
+        match prior.kind() {
+            SyntaxKind::Whitespace => {
+                let count = prior.to_string().chars().filter(|c| *c == '\n').count();
+                if first || !element_spacing {
+                    // first element cannot have a blank line before it
+                    if count > 1 {
+                        state.add(excess_blank_line(prior.text_range().to_span()));
+                    }
+                } else if count < 2 && element_spacing {
```

Comment:
```suggestion
                } else if count < 2 {
```

If the first `if` block didn't catch, `element_spacing` is guaranteed to be `true`.

Reply:
Doesn't `||` short-circuit? So if `first` == `true`, then we have no guarantee about `element_spacing`.

Reply:
Yeah that is true. Revisiting this, I thought I had worked through the logic tree in terms of the caller that it was extraneous. But looking at it again, I can't reach the same conclusion.